### PR TITLE
Bot fill

### PR DESCRIPTION
**Pull request by slipher**
This adds a `bot fill` command to automatically add bots to the team, addressing #522. If the number of players on the team is less than the specified threshold, bots will be added to meet it. If the number of players is higher than the threshold and there are bots which were automatically added, automatically added bots will be removed.

Other changes to bot handling:
 - Bots are always removed after a map restart, even if a keep teams option was selected.
 - Spectate bots vote removed.
 - `bot spec` and `bot unspec` commands removed.

### CODE REVIEW

**Comment by illwieckz** [comment]
Nice, I will try it out!

**Comment by DolceTriade** [comment]
Ugh, this github review interface is garbage. Not yet done reviewing this change.

File: daemon/src/engine/renderer/gl_shader.cpp [hunk 1]
```diff
@@ -1022,8 +1022,8 @@ GLuint GLShaderManager::CompileShader( Str::StringRef programName,
 				       GLenum shaderType ) const
 {
 	GLuint shader = glCreateShader( shaderType );
-	const GLchar *texts[headers.size() + 1];
-	GLint lengths[headers.size() + 1];
+	std::vector<const GLchar*> texts(headers.size() + 1);
```
Comment (DolceTriade) [inline daemon/src/engine/renderer/gl_shader.cpp]: Could you include the rationale for not using VLAs in the commit message?

File: src/sgame/sg_admin.cpp [hunk 2]
```diff
@@ -5579,135 +5579,128 @@ void G_admin_cleanup()
...
+	                                        "            bot names (aliens | humans) <names>…\n"
+	                                        "            bot names (clear | list)" ) );
+	ADMP( bot_usage );
+}
 
-	static const char bot_usage[] = QQ( N_( "^3bot: ^7usage: bot add [^5name|*^7] [^5aliens|humans^7] (^5skill^7) (^5behavior^7)\n"
-	                                        "            bot del [^5name|all^7]\n"
-	                                        "            bot names [^5aliens|humans^7] [^5names…^7]\n"
-	                                        "            bot names [^5clear|list^7]" ) );
+static team_t TeamFromString2( const char* s ) {
```
Comment (DolceTriade) [inline src/sgame/sg_admin.cpp]: This should go in bg_misc.cpp. BG_ParseTeamName or something.

File: src/sgame/sg_admin.cpp [hunk 3]
```diff
@@ -5579,135 +5579,128 @@ void G_admin_cleanup()
...
+	}
+	if ( !Q_stricmp( s, "h" ) || !Q_stricmp( s, "humans" ) )
+	{
+		return team_t::TEAM_HUMANS;
+	}
+	return team_t::TEAM_NONE;
+}
 
-	if ( trap_Argc() < min_args )
+static bool BotAddCmd( gentity_t *ent, const Cmd::Args& args ) {
```
Comment (DolceTriade) [inline src/sgame/sg_admin.cpp]: Please be consistent about where you put the * and &. Current code favors putting the &/* on the variable and not the type.

File: src/sgame/sg_admin.cpp [hunk 4]
```diff
@@ -5579,135 +5579,128 @@ void G_admin_cleanup()
...
-			}
-			else if ( skill_int > 10 )
-			{
-				skill_int = 10; //skill arg too bit, reset to 10
-			}
-		}
+	int skill;
+	if ( args.Argc() >= 5 )
+	{
+		skill = Math::Clamp( atoi( args[4].data() ), 1, 10 );
```
Comment (DolceTriade) [inline src/sgame/sg_admin.cpp]: Perhaps print a warning to the user letting them know that the skill they entered was incorrect?

File: src/sgame/sg_admin.cpp [hunk 5]
```diff
@@ -5579,135 +5579,128 @@ void G_admin_cleanup()
...
+		return BotAddCmd( ent, args );
+	}
+	else if ( !Q_stricmp( arg1, "del" ) && args.Argc() == 3 )
 	{
 		RETURN_IF_INTERMISSION;
 
-		clientNum = G_ClientNumberFromString( name, err, sizeof( err ) );
-		if ( !Q_stricmp( name, "all" ) )
+		char err[MAX_STRING_CHARS];
+		const char* name = args[2].data();
```
Comment (DolceTriade) [inline src/sgame/sg_admin.cpp]: const char *name

File: src/sgame/sg_admin.cpp [hunk 6]
```diff
@@ -5579,135 +5579,128 @@ void G_admin_cleanup()
...
+				return false;
+			}
 			G_BotDel( clientNum ); //delete the bot
 		}
 	}
-	else if ( !Q_stricmp( arg1, "names" ) )
+	else if ( !Q_stricmp( arg1, "names" ) && args.Argc() >= 3 )
 	{
-		if ( !Q_stricmp( name, "humans" ) || !Q_stricmp( name, "h" ) )
+		const char* name = args[2].data();
```
Comment (DolceTriade) [inline src/sgame/sg_admin.cpp]: const char *name

File: src/sgame/sg_admin.cpp [hunk 7]
```diff
@@ -5579,135 +5579,128 @@ void G_admin_cleanup()
...
+		team_t team = TeamFromString2( name );
+
+		if ( team == team_t::TEAM_HUMANS )
 		{
-			i = G_BotAddNames( TEAM_HUMANS, 3, trap_Argc() );
+			int i = G_BotAddNames( TEAM_HUMANS, 3, trap_Argc() );
```
Comment (DolceTriade) [inline src/sgame/sg_admin.cpp]: team_t::TEAM_HUAMNS?

File: src/sgame/sg_admin.cpp [hunk 8]
```diff
@@ -5579,135 +5579,128 @@ void G_admin_cleanup()
...
 			ADMP( va( "%s %d", Quote( P_( "added $1$ human bot name", "added $1$ human bot names", i ) ), i ) );
 		}
-		else if ( !Q_stricmp( name, "aliens" ) || !Q_stricmp( name, "a" ) )
+		else if ( team == team_t::TEAM_ALIENS )
 		{
-			i = G_BotAddNames( TEAM_ALIENS, 3, trap_Argc() );
+			int i = G_BotAddNames( TEAM_ALIENS, 3, trap_Argc() );
```
Comment (DolceTriade) [inline src/sgame/sg_admin.cpp]: team_t::TEAM_ALIENS?

**Comment by illwieckz** [comment]
So, when I use the command `bot fill N` or `bot fill N team` the team are filled as expected, but when I join a team, all the bots from my team are removed then readded (minus one) so the bot count is ok, but it does not happen the right way. :D In most of the case, some bots continue to weirdly reconnect/disconnect infinitely while I'm playing. It looks like the excluded bot try to re-enter the game and is kicked automatically, then retry etc.

File: src/sgame/sg_main.cpp [hunk 9]
```diff
@@ -926,6 +926,7 @@ G_ShutdownGame
 */
 void G_ShutdownGame( int restart )
```
Comment (DolceTriade) [inline src/sgame/sg_main.cpp]: void G_ShutdownGame( int /* restart */ ) ? Then you can delete the Q_UNUSED below.

File: src/sgame/sg_bot.h [hunk 10]
```diff
@@ -87,7 +87,11 @@ typedef struct
 	botNavCmd_t nav;
 } botMemory_t;
 
-bool G_BotAdd( const char *name, team_t team, int skill, const char* behavior );
+constexpr int BOT_DEFAULT_SKILL = 5;
+#define BOT_DEFAULT_BEHAVIOR "default"
```
Comment (DolceTriade) [inline src/sgame/sg_bot.h]: const char BOT_DEFAULT_BEHAVIOR[] = "default";

File: src/sgame/sg_bot.h [hunk 11]
```diff
@@ -87,7 +87,11 @@ typedef struct
...
+#define BOT_NAME_FROM_LIST "*"
```
Comment (DolceTriade) [inline src/sgame/sg_bot.h]: const char BOT_NAME_FROM_LIST[] = "*";

File: src/sgame/sg_bot.h [hunk 12]
```diff
@@ -87,7 +87,11 @@ typedef struct
...
+
+bool G_BotAdd( const char *name, team_t team, int skill, const char* behavior, bool filler = false );
```
Comment (DolceTriade) [inline src/sgame/sg_bot.h]: const char *behavior

**Comment by slipher** [comment]
I've fixed the bug reported by @illwieckz. When there were too many bots on a team, I forgot to check that the bots to be deleted are on that same team :)

**Comment by slipher** [comment]
I have addressed the review comments.

**Comment by illwieckz** [comment]
@slipher I found a way to retrigger the stuck bot in reentering loop anyway:

```
/devmap parpax
/bot fill 5
/team humans
/team aliens
/bot fill 2 humans
/bot fill 1 aliens
/team humans
```

The alien bot will fail to join the alien team.

Wait for the map to be loaded and wait for the bot to join/quit between each command.

File: src/sgame/sg_admin.cpp [hunk 13]
```diff
@@ -5579,203 +5579,169 @@ void G_admin_cleanup()
...
+	{
+		BotUsage( ent );
+		return false;
+	}
+	int count = atoi( args[2].data() );
+	std::vector<team_t> teams;
+	if ( args.Argc() >= 4 )
+	{
+		team_t team = BG_PlayableTeamFromString(args[3].data());
+		if (team == TEAM_NONE)
```
Comment (DolceTriade) [inline src/sgame/sg_admin.cpp]: team_t::

File: src/sgame/sg_admin.cpp [hunk 14]
```diff
@@ -5579,203 +5579,169 @@ void G_admin_cleanup()
...
-		{
-			G_BotDel( clientNum ); //delete the bot
-		}
+		teams = { team };
 	}
-	else if ( !Q_stricmp( arg1, "spec" ) )
+	else
 	{
-		RETURN_IF_INTERMISSION;
+		teams = { TEAM_ALIENS, TEAM_HUMANS };
```
Comment (DolceTriade) [inline src/sgame/sg_admin.cpp]: team_t::

**Pull request merged**